cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/UserTaskManager.java

```diff
@@ -317,38 +324,56 @@ private synchronized void removeFromPurgatory(UserTaskInfo userTaskInfo) {
-        iterator.remove();
+    for (RetentionType retentionType: RetentionType.cachedValues()) {
+      long completedUserTaskRetentionTimeMs = (retentionType == ADMIN_CRUISE_CONTROL || retentionType == ADMIN_KAFKA) ?
+                                              _completedAdminUserTaskRetentionTimeMs : _completedMonitorUserTaskRetentionTimeMs;
+      for (Iterator<Map.Entry<UUID, UserTaskInfo>> iterator =
+          _uuidToCompletedUserTaskInfoMap.get(retentionType).entrySet().iterator(); iterator.hasNext(); ) {
+        Map.Entry<UUID, UserTaskInfo> entry = iterator.next();
+        if (entry.getValue().startMs() + completedUserTaskRetentionTimeMs < _time.milliseconds()) {
+          removeFromPurgatory(entry.getValue());
+          iterator.remove();
+        }
       }
     }
   }
 
   /**
-   * Allow other subsystems to retrieve {@link UserTaskInfo} using uuid. This method assumes uuid comes from some internal
-   * bookkeeping state, not an external http request. Thus it does not check request parameters like
-   * {@link UserTaskManager#getUserTaskByUserTaskId} does.
-   * @param uuid UUID associated with a user request or execution.
-   * @return {@link UserTaskInfo} which contains information about a user request.
+   * Mark the task which is currently executed in {@link com.linkedin.kafka.cruisecontrol.executor.Executor}.
+   *
+   * @param uuid UUID associated with the task in-execution.
+   * @return {@link UserTaskInfo} associated with the task in-execution.
    */
-  public synchronized UserTaskInfo getUserTaskById(String uuid) {
+  public synchronized UserTaskInfo markTaskInExecution(String uuid) {
     if (uuid == null) {
       return null;
     }
 
     UUID userTaskId = UUID.fromString(uuid);
-    if (_allUuidToUserTaskInfoMap.get(TaskState.COMPLETED).containsKey(userTaskId)) {
-      return _allUuidToUserTaskInfoMap.get(TaskState.COMPLETED).get(userTaskId);
+    for (Map<UUID, UserTaskInfo> infoMap : _uuidToCompletedUserTaskInfoMap.values()) {
+      if (infoMap.containsKey(userTaskId)) {
+        _inExecutionUserTaskInfo = infoMap.remove(userTaskId).setState(TaskState.IN_EXECUTION);
+      }
     }
 
-    if (_allUuidToUserTaskInfoMap.get(TaskState.ACTIVE).containsKey(userTaskId)) {
-      return _allUuidToUserTaskInfoMap.get(TaskState.ACTIVE).get(userTaskId);
+    if (_uuidToActiveUserTaskInfoMap.containsKey(userTaskId)) {
+      _inExecutionUserTaskInfo = _uuidToActiveUserTaskInfoMap.remove(userTaskId).setState(TaskState.IN_EXECUTION);
     }
 
-    return null;
+    return _inExecutionUserTaskInfo;
+  }
+
+  /**
+   * Unmark the task in-execution once {@link com.linkedin.kafka.cruisecontrol.executor.Executor} has finish the execution.
+   *
+   * @param uuid UUID associated with the task in-execution.
+   */
+  public synchronized void markTaskFinishExecution(String uuid) {
+    if (_inExecutionUserTaskInfo == null || _inExecutionUserTaskInfo.userTaskId() != UUID.fromString(uuid)) {
```

Comment:
Incorrect comparison of UUIDs -- I believe we intend to use `_inExecutionUserTaskInfo.userTaskId().equals(UUID.fromString(uuid))`?

Please see https://stackoverflow.com/questions/23589058/to-compare-uuid-can-i-use-or-have-to-use-uuid-equalsuuid